java/src/main/java/org/rocksdb/RocksIterator.java

```diff
@@ -37,6 +39,28 @@ protected RocksIterator(RocksDB rocksDB, long nativeHandle) {
     return key0(nativeHandle_);
   }
 
+  /**
+   * <p>Return the key for the current entry.  The underlying storage for
+   * the returned slice is valid only until the next modification of
+   * the iterator.</p>
+   *
+   * <p>REQUIRES: {@link #isValid()}</p>
+   *
+   * @param key the out-value to receive the retrieved key.
+   *     It is using position and limit. Limit is set according to key size.
+   *     Supports direct buffer only.
+   * @return The size of the actual key. If the return key is greater than the
+   *     length of {@code key}, then it indicates that the size of the
+   *     input buffer {@code key} is insufficient and partial result will
+   *     be returned.
+   */
+  public int key(ByteBuffer key) {
```

Comment:
Hmm. I don't like that this `key` method side-effects the parameter, while the other is in essence a simple getter. Can this be called `copyKey` or maybe `xferKey` or `readKey` or `saveKey`?

(`get` doesn't bother me in this way, as its broad semantic meaning is more clear.)

Reply:
@pdillinger I have to say I really dislike that idea. It is totally normal for ByteBuffer to be modified. In fact the whole point of this method is to modify the byte-buffer. In Java any function that takes a ByteBuffer as a parameter, for reads or writes will modify the state of the ByteBuffer - this is a very normal convention for Java.

I also think it is important to maintain consistency in the API, this `key(ByteBuffer)` method, mirrors the existing `byte[] key()` method.

Reply:
Someone sees this reading code: `iter.key(k)`

Does this set something or get something? Taking a ByteBuffer is not a clue in either direction. And since the method name is not a verb (e.g. Google java style https://google.github.io/styleguide/javaguide.html#s5.2.3-method-names, or Java Beans style getter https://en.wikipedia.org/wiki/JavaBeans) I don't consider it obvious to a general Java audience.

However, I do see that in our Java codebase, we have a consistent style that noun method names are getters. For better or worse, this is consistent, and good enough for me.